test/functional/test_framework/p2p.py

```diff
@@ -598,6 +598,11 @@ def wait_for_disconnect(self, timeout=60):
         test_function = lambda: not self.is_connected
         self.wait_until(test_function, timeout=timeout, check_connected=False)
 
+    def wait_for_reconnect(self, timeout=60):
```

Comment:
ideally, first `wait_for_connect` waits on v2 connection and the `wait_for_connect` a few lines below waits for v1 connection.

it's possible that the v2 reconnection happens quickly and the first `wait_for_connect` detects a v2 connection directly. Then we'd be stuck waiting for disconnect and connect (in the lines below) to happen (even though it has already happened).

Reply:
Maybe it would work just not checking for the racy intermediate states, but instead waiting until we have received the "version" message via v1? Something like

 ```
def test_function():
     if not (self.is_connected and self.last_message.get('version') and self.v2_state is None):
         return False
     return True
self.wait_until(test_function, timeout=timeout, check_connected=False  )
```

Reply:
good idea! that should work.